jdaviz/configs/mosviz/plugins/parsers.py

```diff
@@ -755,29 +755,32 @@ def mos_niriss_parser(app, data_dir):
             print(f"Loading: {flabel} sources")
 
             with fits.open(fname, memmap=False) as temp:
-                # treat all HDUs missing SRCTYPE attribute as extended sources
-                # TODO: Remove this once valid SRCTYPE values are present in all headers
-                for hdu in temp:
-                    if ("SRCTYPE" in hdu.header
-                            and (hdu.header["SRCTYPE"] in ("POINT", "EXTENDED"))):
-                        pass
-                    else:
-                        hdu.header["SRCTYPE"] = "EXTENDED"
-
-                # read all HDUs with SpectrumList, then only keep those that
-                # correspond with sources in catalog
-                # (read() is slow... would also LOVE to do this in one step!)
-                specs = SpectrumList.read(temp, format="JWST x1d multi")
-                specs_cut = [sp for sp in specs if sp.meta['header']['SOURCEID']
-                             in cat_id_dict.keys()]
+                # Filter out HDUs we care about
+                filtered_hdul = fits.HDUList([hdu for hdu in temp if (
+                    (hdu.name in ('PRIMARY', 'ASDF')) or
+                    (hdu.header.get('SOURCEID', None) in cat_id_dict.keys()))])
```

Comment:
`cat_id_dict.keys()` can be cached first outside of the list comprehension, though not sure how much speed up you get but something to think about.

Reply:
Implemented in https://github.com/spacetelescope/jdaviz/pull/1696/commits/ebc2cc310e4807fe1ebf747e59a648a9f7ee5bec; thanks for the suggestion!